networkx/algorithms/approximation/distance_measures.py

```diff
@@ -0,0 +1,138 @@
+    forward and a backward BFS. Let $a_1$ and $a_2$ be the farthest nodes in the forward and
+    backward cases, respectively. Then, it computes the backward eccentricity of $a_1$ using
+    a backward BFS and the forward eccentricity of $a_2$ using a forward BFS.
+    Finally, it returns the best lower bound between the two.
+
+    In both cases, the time complexity is linear with respect to the size of G.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    seed : integer, random_state, or None (default)
+        Indicator of random number generation state.
+        See :ref:`Randomness<randomness>`.
+
+    Returns
+    -------
+    d : integer
+       Lower Bound on the Diameter of G
+
+    Raises
+    ------
+    NetworkXError
+        If the graph is empty or
+        If the graph is undirected and not connected or
+        If the graph is directed and not strongly connected.
+
+    See Also
+    --------
+    networkx.algorithms.distance_measures.diameter
+
+    References
+    ----------
+    .. [1] Magnien, Clémence, Matthieu Latapy, and Michel Habib.
+       *Fast computation of empirically tight bounds for the diameter of massive graphs.*
+       Journal of Experimental Algorithmics (JEA), 2009.
+       https://arxiv.org/pdf/0904.2728.pdf
+    .. [2] Crescenzi, Pierluigi, Roberto Grossi, Leonardo Lanzi, and Andrea Marino.
+       *On computing the diameter of real-world directed (weighted) graphs.*
+       International Symposium on Experimental Algorithms. Springer, Berlin, Heidelberg, 2012.
+       https://courses.cs.ut.ee/MTAT.03.238/2014_fall/uploads/Main/diameter.pdf
+    """
+    # if G is empty
+    if not G:
+        raise nx.NetworkXError("Expected non-empty NetworkX graph!")
+    # if there's only a node
+    if G.number_of_nodes() == 1:
+        return 0
+    # if G is directed
+    if G.is_directed():
+        if not nx.is_strongly_connected(G):
+            raise nx.NetworkXError("DiGraph not strongly connected.")
+        return _two_sweep_directed(G, seed)
+    # else if G is undirected
+    if not nx.is_connected(G):
+        raise nx.NetworkXError("Graph not connected.")
+    return _two_sweep_undirected(G, seed)
+
+
+def _two_sweep_undirected(G, seed=None):
+    """Helper function for finding a lower bound on the diameter
+        for undirected Graphs.
+
+        The idea is to pick the farthest node from a random node
+        and return its eccentricity.
+
+        ``G`` is a NetworkX undirected graph.
+
+    .. note::
+
+        ``seed`` is a random.Random or numpy.random.RandomState instance
+    """
+    # select a random source node
+    source = seed.sample(G.nodes(), 1)[0]
+    # take a node v that is (one of) the farthest nodes from the source
+    *_, (_, v) = nx.bfs_edges(G, source)
+    # compute the distances from v to the other nodes
+    distances = nx.single_source_shortest_path_length(G, v)
+    # return the eccentricity of v
+    return max(distances.values())
```

Comment:
There is already an eccentricity function in networkx. Can we use that?
And `seed.choice` seems more appropriate than `sample` when selecting one node.
Finally, `shortest_path_length` has some advantages over `bfs_edges` in terms of unpacking the result to get a node. It is also faster for all graphs I tried (from lattice to barabasi-albert to binomial_graph) but about 10 times faster for the binomial.

```suggestion
    # select a random source node
    source = seed.choice(G)
    # take a node that is (one of) the farthest nodes from the source
    *_, node = nx.shortest_path_length(G, source) 
    return nx.eccentricity(G, node)
```

Similarly for directed only do it twice, once for G and once for G_reversed and take max of the two.